ASoC: Intel: Add support for rt1015_rt5682

yongzhi1 opened the pull request:
This patch adds sof_rt1015_rt5682 machine driver.

Tested both codecs on JasperLake platform individually.

Signed-off-by: Yong Zhi <yong.zhi@intel.com>

lyakh commented on sound/soc/intel/boards/sof_rt1015_rt5682.c:
```diff
@@ -0,0 +1,386 @@
+static int card_late_probe(struct snd_soc_card *card)
+{
+	struct card_private *ctx = snd_soc_card_get_drvdata(card);
+	struct snd_soc_acpi_mach *mach = (card->dev)->platform_data;
```
superfluous parentheses

yongzhi1 replied:
Ack.

lyakh commented on sound/soc/intel/boards/sof_rt1015_rt5682.c:
```diff
@@ -0,0 +1,386 @@
+{
+	struct card_private *ctx = snd_soc_card_get_drvdata(card);
+	struct snd_soc_acpi_mach *mach = (card->dev)->platform_data;
+	struct snd_soc_component *component = NULL;
```
superfluous initialisation

yongzhi1 replied:
Ack.

lyakh commented on sound/soc/intel/boards/sof_rt1015_rt5682.c:
```diff
@@ -0,0 +1,386 @@
+	DAILINK_COMP_ARRAY(COMP_CODEC("dmic-codec", "dmic-hifi")));
+
+SND_SOC_DAILINK_DEF(dmic16k_pin,
+	DAILINK_COMP_ARRAY(COMP_CPU("DMIC16k Pin")));
```
I was told we don't use "dmic01" / dmic16k" names any more, does that only hold for topology files or for machine drivers too?

yongzhi1 replied:
Not sure here either.

ranj063 replied:
@lyakh these are the CPU DAI names defined in hda-dai.c. The name is correct here.

lyakh commented on sound/soc/intel/boards/sof_rt1015_rt5682.c:
```diff
@@ -0,0 +1,386 @@
+	card->dev = &pdev->dev;
+
+	/* override plaform name, if required */
+	mach = (&pdev->dev)->platform_data;
```
you meant to write `pdev->dev.platform_data`

yongzhi1 replied:
Ok.

lyakh commented on sound/soc/intel/common/soc-acpi-intel-jsl-match.c:
```diff
@@ -2,7 +2,7 @@
  * soc-apci-intel-jsl-match.c - tables and support for JSL ACPI enumeration.
  *
- * Copyright (c) 2019, Intel Corporation.
+ * Copyright (c) 2020, Intel Corporation.
```
shouldn't this be 2019-2020?

yongzhi1 replied:
Ack.

plbossart commented:
@yongzhi1 could we look at re-using sof_rt5682? we already have a case where we use the maxim amp, this would be another quirk to use a different amp for jsl only.

yongzhi1 commented:
Hi, @plbossart , I also think of reuse of the sof_rt5682, the only drawback is that validating code changes on all platforms will likely become a development bottleneck.

plbossart commented:
> Hi, @plbossart , I also think of reuse of the sof_rt5682, the only drawback is that validating code changes on all platforms will likely become a development bottleneck.

well you can also look at the flip side that the code will be updated faster since it's used by more people :-)
If you only add a tiny part related to a relatively contained part it's probably better to reuse rather than redevelop an entire machine driver.

yongzhi1 commented:
> > Hi, @plbossart , I also think of reuse of the sof_rt5682, the only drawback is that validating code changes on all platforms will likely become a development bottleneck.
> 
> well you can also look at the flip side that the code will be updated faster since it's used by more people :-)
> If you only add a tiny part related to a relatively contained part it's probably better to reuse rather than redevelop an entire machine driver.

In this case, can I rewrite a shared version for you to take a look?

plbossart commented:
> > > Hi, @plbossart , I also think of reuse of the sof_rt5682, the only drawback is that validating code changes on all platforms will likely become a development bottleneck.
> > 
> > 
> > well you can also look at the flip side that the code will be updated faster since it's used by more people :-)
> > If you only add a tiny part related to a relatively contained part it's probably better to reuse rather than redevelop an entire machine driver.
> 
> In this case, can I rewrite a shared version for you to take a look?

sure, if you follow my ideas of course I'll take a look :-)

yongzhi1 commented:
> > > > Hi, @plbossart , I also think of reuse of the sof_rt5682, the only drawback is that validating code changes on all platforms will likely become a development bottleneck.
> > > 
> > > 
> > > well you can also look at the flip side that the code will be updated faster since it's used by more people :-)
> > > If you only add a tiny part related to a relatively contained part it's probably better to reuse rather than redevelop an entire machine driver.
> > 
> > 
> > In this case, can I rewrite a shared version for you to take a look?
> 
> sure, if you follow my ideas of course I'll take a look :-)

Thanks, updated  ef8368e, have not done much test yet, just want to make sure the overall direction is correct.

ranj063 commented on sound/soc/intel/boards/sof_rt5682.c:
```diff
@@ -417,6 +481,17 @@ static struct snd_soc_dai_link_component max98357a_component[] = {
+		.name = "i2c-10EC1015:01",
+		.dai_name = "rt1015-aif",
+	},
+};
```
@yongzhi1 can we please use the macro SND_SOC_DAILINK_DEF here for defining the codec components?

yongzhi1 replied:
Principally yes, I also tried to match the original style here :)

kv2019i replied:
@ranj063 @yongzhi1 That's probably better done in a separate patch if you need to touch existing definitions.

ranj063 replied:
ack!

ranj063 commented on sound/soc/intel/boards/sof_rt5682.c:
```diff
@@ -669,6 +751,11 @@ static int sof_audio_probe(struct platform_device *pdev)
 
 	sof_audio_card_rt5682.dai_link = dai_links;
 
+	if (!strcmp(pdev->name, "jsl_rt1015_rt5682")) {
```
@yongzhi1 do you think it makes sense to add a quirk for the AMP type rt1015/max for this change instead of using the driver name? It will make reusing the machine driver for a different platform with the same codec combination easier

yongzhi1 replied:
Yes, the driver is for Google waddle-dee, but the DMI data(product name etc.) is not yet finalized to use quirk cb.

ranj063 replied:
but you could also do the same in the driver data no?

yongzhi1 replied:
yes, you are right, I can define a new BIT for the rt1015, and use driver_data to set it.

yongzhi1 commented:
Updated https://github.com/yongzhi1/linux/commit/76def42ea21697831003096108d655e86e881c7b
For Ver2:
- Added SOF_RT1015_SPEAKER_AMP_PRESENT quirk bit.
- Updated snd_soc_acpi_intel_jsl_machines with spk quirk to make SOF_SPEAKER_AMP_PRESENT real.

lyakh commented on sound/soc/intel/boards/sof_rt5682.c:
```diff
@@ -260,6 +262,42 @@ static struct snd_soc_ops sof_rt5682_ops = {
+		}
+	}
+
+	return ret;
```
I often enough have to search through nested function calls to find out what specific values a function can return, e.g. only 0 and negative errors, or also positive values. So, it helps when functions are explicit about that. Here it *seems* (unless `snd_soc_dai_set_pll()` or `snd_soc_dai_set_sysclk()` can also return positive values), that this function should only return 0 or negative errors. If this is indeed the case, you can make this last statement `return 0;`. If you do that you can also remove initialisation of `ret`.

yongzhi1 replied:
Ack, thanks for the review.

ranj063 commented on sound/soc/intel/boards/sof_rt5682.c:
```diff
@@ -1,5 +1,5 @@
+// Copyright(c) 2019-2020 Intel Corporation.
 
 /*
  * Intel SOF Machine Driver with Realtek rt5682 Codec
```
@yongzhi1 can you please also fix the comment on the next line to add that it supports RT1015 too?

yongzhi1 replied:
Ack, will update in next version.

yongzhi1 commented:
Updated https://github.com/thesofproject/linux/commit/4bff90d67f523610fcc7d614083944e93b2911c7

For Ver 3:
- Removed ret init, and changed ret to 0 at sof_rt1015_hw_params() exit.
- Updated file comments to include RT1015 spk amp.

yongzhi1 commented:
Patch rebased: https://github.com/thesofproject/linux/commit/1bf657481164955286585764693f26e6658ffc46
